### ISSUE TITLE
`groupby.transform` failing on dataframe with duplicates in column

### ISSUE THREAD

**Issue by guidocioni** [dask/dask#8800]
I'm trying to process parquet files stored in AWS S3.

The files are read simply with
```python
with Client(n_workers=6) as client:
    df = dd.read_parquet('s3://lightnings_*.gzip.parquet')
    df['time_plot'] = df.time.dt.round(freq='15 min')
```
It looks like this 

| time                             |   latitude |   longitude | time_plot                 |
|:---------------------------------|-----------:|------------:|:--------------------------|
| 2021-04-01 00:00:30.688104+00:00 |    36.4179 |     34.1496 | 2021-04-01 00:00:00+00:00 |
| 2021-04-01 00:25:24.591956+00:00 |    37.7171 |    -10.604  | 2021-04-01 00:30:00+00:00 |

There are some duplicated rows (not shown here)!

When trying to apply a `groupby` operation

```python
    df['seconds_elapsed'] = df.groupby("time_plot").time.transform(
            lambda x: (x - x.min()).dt.seconds)
```

I get an error: `ValueError: cannot reindex on an axis with duplicate labels`.
The same code runs fine on a normal `pandas.DataFrame` by calling `.compute()` before. 

As I thought the problem was related to the duplicates I tried to remove them before calling the `groupby.transform`, i.e.

```python
    df = df.drop_duplicates(subset=['time'], split_out=6)
```

This takes a long time, uses a lot of memory and still does not fix the problem as I still get the error reported before. Note that I had to use `split_out` otherwise the workers were running out of memory. 

As workaround I'm currently loading the dataset into memory (`df.compute()`) before passing it to the `groupby` operations, but this is quite slow as this is applied serially so it would be good to adapt the workflow with Dask. 

Anyone has suggestions on what I'm doing wrong?

**Comment by guidocioni** [comment]
Yeah that's right.
I have to say the assignment of a groupby.apply operation to the original dataframe often causes issues...
Any idea why you don't get this error if converting the dask dataframe to a pandas dataframe before? The index should be the same...right?

**Comment by jsignell** [comment]
> Yeah that's right. I have to say the assignment of a groupby.apply operation to the original dataframe often causes issues... Any idea why you don't get this error if converting the dask dataframe to a pandas dataframe before? The index should be the same...right?

Yeah I think in pandas the data doesn't really get shuffled, so when assigning the output to the column in the dataframe the indices match. For dask on the other hand, the indices don't match so that's when the reindex happens and then it throws an error because the index values are non-unique. 

That's my current understanding anyways. So I think a global monotonically increasing index would resolve this. That was one of the suggestions in the issues I linked.
